### DISCUSSION TITLE
Session networking handling rewrite testing

### DISCUSSION THREAD

**Discussion by Frooxius** [Yellow-Dog-Man/Resonite-Issues#3021] [Pre-release Testing]
Hello everyone!

I've got another thing that needs some testing before it gets released to public, since it can cause some stuff to blow up.

## Build Version
2024.9.28.1132

## What is being tested
The session networking handling system has been completely restructured and rewritten as part of performance optimizations, switching over to the Task Parallel Library for a lot of the heavy duty processing instead of manual thread management. More of the processing is now parallelized as a result and also hoisted out of the world update loop. There are some small optimizations and improvements sprinkled along as well.

The performance goal of this change is to eliminate the server hitches when new users join, especially populated sessions (**IMPORTANT:** This doesn't affect hitches from loading assets, that's a separate future set of optimizations), which is documented in this issue: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1477

Another goal of this is to make the code and its flow much more readable and manageable, which will make a lot of things easier in the future: fixing bugs, improving/expanding security and privacy and implementing new features like mesh networking.

The system that was rewritten is responsible for a number of things:
- Initial handshake process when joining a session, which determines if the user is allowed into the session or not (this includes authentication of the user and host)
- Encoding/decoding of sync messages and queuing them for network transmission
- Routing sync messages to appropriate systems for decoding and applying (this includes full and delta changes, streams and control messages)
- Managing connections to the session - e.g. triggering world to close or users to be removed when connections close
- Session asset transfer - the system itself wasn't changed much yet, though the routing and processing of messages for it was, which is now hoisted out of the world update loop

Since this is a critical system for Resonite and being able to join, host and exist in sessions, I'd like this to be tested before it's fully deployed, since some stuff can break. I did some preliminary smoke testing, but to fully test the system it needs a number of users and networking scenarios.

## How to test
- On Steam, switch to the `prerelease` branch and make sure the build updates before you run it
- Make sure your build version matches the one at the top of this message

**You will need to test with multiple users** since this affects session hosting and joining sessions.

Here's a list of things you should try out (or at least some of these) and focus on while testing:

### What to test
- Hosting sessions (both client and headless)
- Joining sessions
- Different session accessibility settings - public, contacts, contacts plus, private, hidden...
- Test scenarios which reject users from being able to join - e.g. when the session is full or when you ban users
- Custom join verifier (this was touched with the rewrite)
- Kicking users from the session
- Import assets into the session to make sure asset transfers work fine
- Try different protocols if you can - LNL, Steam Networking Sockets, TCP
- Streaming audio into the session
- Generally using the session as usual - play around, do stuff you normally do to make sure it doesn't explode

### What to watch out for
These are good things to watch out for when testing:

- Sessions exploding and crashing - probably most obvious
- Issues joining a session when you normally should be able to
- The world syncing breaking - things stop updating or get corrupted
- Streams (poses, voice, audio streams) breaking in ways that are not present on the mail build
- Other oddities or behaviors not present on the main build - if you notice something not listed here and you're not sure, feel free to bring it up! At worst we'll redirect it somewhere else.

## What NOT to test & report
Generally things that do not involve this system - any existing unrelated bugs with components and other systems.
- Session asset transfers occasionally getting stuck with lots of users - this is a separate bug: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2653
-- **IMPORTANT:** If you can get it to happen a lot more frequently than it does usually, then DO report it.
- Hitching from users loading in their avatars - there's multiple sources of hitches when user joins. One is client-side when the avatar loads - this is not affected by this rework and will still happen - it'll depend on the complexity of the avatar
- Queued packets - if you start queuing, this is a separate issue, though shouldn't happen as much. Try increasing the LNL Window Size
- If a component or other system is behaving oddly and you're unsure, test it on the current public build. If the issue is present there as well, don't report it as part of this testing
- If the session explodes (e.g. the world closes or you crash) - you can report this immediately, even if you're not aware of the reproduction steps - make sure to **include logs!**

## If you find something that got broken
- Create a GitHub issue for a pre-release bug: https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/new?assignees=shiftyscales&labels=prerelease&projects=&template=prerelease.yml
- Make sure to include the log
- Please include description of what happened. Be succinct, but exact.

## If you find that stuff works okay
Please report it too and include details on what you generally tested, so we know things are okay. You don't need to make a new issue for this - just a commend in this Discussion is fine.

This lets us know that a bunch of stuff was tested and get better confidence that stuff won't break with release.

## How long will testing go for?
With the nature of these changes issues should become apparent pretty quickly - at least the frequent ones. If we get a bunch of reports of things working okay, we might release this on Monday already. If there's blocker issues cropping up, we'll hold up the release until things are resolved.

**Comment by j4lc** [comment]
When trying to update my mod for the prerelease, I see that things like `world.Session.TotalReceivedConfirmations` are not found anymore and not under `world.Session.Sync`.

Where some of those metrics deprecated with the change?

**Reply by Frooxius** [reply]
It's under `world.Session.Messages.Incoming`.

I didn't remove any stats, but they're in different places now.

**Reply by j4lc** [reply]
I see, thanks!

Also tested with a small session (4-5 users max) and only one issue: for some reason someone couldn't join the world at the first try.

I'll try to create a setup to reproduce that.

Otherwise, no apparent issues.

**Comment by orange3134** [comment]
This is a great update!
I tested it in a session of 8-10 people and there was no voice interruption when someone joined the world.

The sync of values still seems to be paused, is this unavoidable?

https://github.com/user-attachments/assets/1be9d7ee-87c1-40f5-9a90-722a36a777ce

This is an item that visualizes the delay by having each user update their line position every frame, but we found that the updates stopped for a while after the user joined the world.

I have tried several games and items for about an hour and so far have not found any glitches.

**Reply by Frooxius** [reply]
Thank you for the testing! That is very cool testing tool as well!

This should mainly help with voice & pose streams, as those can be forwarded outside of the normal game loop. 

The data model values that are not using streams can still stutter during the join due to the capture of the state of the data model, but the encoding pause should be eliminated. It might be what is happening here. I can add a bit of diagnostics to capture how long the capture takes.

There might be still some issue with blocking the transmission queue which could be further optimized. I could potentially split the processing there to happen per-connection.

Was this hosted on a headless?

**Reply by orange3134** [reply]
I see, so it is taking time to capture the data model values.
It would be great if those values could also be synchronized without interruption.

This session was hosted on a regular client, not headless.

**Reply by Frooxius** [reply]
The actual capture of the values needs to be optimized differently if it's the actual cause though separate optimizations, but the cause also might be in the network transmission processing.

I will add a diagnostic in upcoming build to help determine if this is indeed an issue.

**Reply by Readun** [reply]
Cool to see even the Train in the Background stopping aswell.
It might be a good value in getting it smooth too, for all the possible games that can be made. (Racing, shooters and so on)

**Reply by Frooxius** [reply]
It depends on the cause. If this is happening due to network transmission queue, we can fix that as part of this update.

If it's due to the data model capture, that's very different optimization and won't happen as part of this.

Just wait for the build with the diagnostic please.

**Reply by Frooxius** [reply]
@orange3134 I've pushed a new version 2024.9.28.1132

The host will now log the time it takes in the synchronous loop to capture the data model. Can you replicate this issue with this build please?

Once replicated, check the log on the host for following lines: `World initial state captured in:`. Find one that correlates with the user joining in your test (you can press F3 before the user joins on the host to put a marker in the log file).

**Reply by orange3134** [reply]
I will try again with the new version. All my friends have already gone to bed tonight, so I will try again tomorrow.

**Reply by Frooxius** [reply]
No worries, there's no rush!

Even if this turns out to be fixable with this new system, I might not implement this optimization a build until after this part is released anyway.

Thank you for testing!

**Reply by orange3134** [reply]
@Frooxius 
Tested with version 2024.9.28.1132.
[MENTAIKO - 2024.9.28.1132 - 2024-09-29 11_57_20.log](https://github.com/user-attachments/files/17177755/MENTAIKO.-.2024.9.28.1132.-.2024-09-29.11_57_20.log)
The value of World initial state captured in: was less than 1 second with few objects in the world, but increased to 1-3 seconds as the world became more scattered.

From the host's perspective, everything appears to stop at that moment. (It appears the game drawing frames are not being updated)
https://github.com/user-attachments/assets/368fce86-16ee-4502-b483-d8b122b834de

From the non-host's perspective, it appeared that the synchronization of values other than my own temporarily stopped.
https://github.com/user-attachments/assets/2ebf731e-d1dd-4556-872a-8aaa49d525d5

**Reply by Frooxius** [reply]
Thank you for the testing. Since the hitch seems to be caused by the initial capture, that is unfortunatelly not solvable as part of this specific optimization.

It might be worthwhile to test similar scenario with a .NET 8 headless, since the increased runtime performance could make this much faster.

We'll have to evaluate approaches to speed up this initial capture in the future - there are some possibilities, like optimizing the capture itself or maintaining an incrementally updated full state batch.

However with this optimization, the async pose and voice streams are no longer interrupted, which should make the experience much smoother.

The only thing that hitches will be the data model updates.

**Comment by Frooxius** [comment]
I've updated the pre-release with new diagnostic

**2024.9.28.1132**
- Add logging of how long capture of world's initial takes to help diagnose performance (based on feedback by @orange3134)

**Comment by KyuubiYoru** [comment]
Audio stream had almost constant 3% Packet Loss.
![image](https://github.com/user-attachments/assets/7cef40dd-6479-407e-9711-ab2c9b945b2a)

**Reply by Frooxius** [reply]
Does this not happen on the normal build? This can be normal behavior depending on the network used - the changes don't really change networking that much, though there can be more packets being thrown out than normal in some cases.

**Comment by Foxxboxx** [comment]
The Grand Oasis conducted a test on the pre-release. We didn't notice anything particularly out of the ordinary. Everything we tested seemed to work fine outside of already known issues with the main build. 

Thing's we've tested: 

- Headless session
- Headless as a docker container
- Kicking users via client
- Kicking users via console command
- Auto AFK Kicking
- Banning users using the cloud variables
- Access Denied Message using Headless configs
- Pasting from clipboard (images, links, audio clips)
- Playing Youtube Videos
- Streaming music using opus audio streamer
- Connecting usring Relays
- Connecting using Steam Sockets
- Connecting using TCP
- User limit reached

Issues we had: 

- No denied message with cloudvar bans (potentially related to issued 3000)
- Defhammer could not force change to Steam Socket or TCP. NetworkStatistics node always readed LNL (could have been an issue with the node)

Bellow is the log from our headless. 
[headless - 2024.9.28.871 - 2024-09-28 17_13_52.log](https://github.com/user-attachments/files/17176611/headless.-.2024.9.28.871.-.2024-09-28.17_13_52.log)

**Reply by Frooxius** [reply]
Awesome, thank you so much for the thorough testing! 

- The denied message is probably caused by https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/3000 Does this happen on the normal build too?
- Does the Steam Socket / TCP happen with normal build too? From the log it looks like Steam Sockets are not available on the headless, so that could be the culprit. TCP will generally only work on LAN or exposed public IP's if you connect to a direct IP, so that one might be difficult to trigger

**Reply by Foxxboxx** [reply]
Yes, I believe you are right. The headless cannot use Steam Sockets in this instance (Dunno why I didn't think of that lol) The TCP I will test later via LAN. The denied message I can test later tonight when the Oasis has their event to verify its part of issue 3000.

**Reply by bredo228** [reply]
> Yes, I believe you are right. The headless cannot use Steam Sockets in this instance (Dunno why I didn't think of that lol) The TCP I will test later via LAN. The denied message I can test later tonight when the Oasis has their event to verify its part of issue 3000.

Tested with Steam sockets on a Windows Headless, they're working as expected.

**Reply by Frooxius** [reply]
Thanks for the info and testing!

**Reply by Foxxboxx** [reply]
> Awesome, thank you so much for the thorough testing!
> 
> * The denied message is probably caused by [Loading indicator doesn't reveal error message #3000](https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/3000) Does this happen on the normal build too?
> * Does the Steam Socket / TCP happen with normal build too? From the log it looks like Steam Sockets are not available on the headless, so that could be the culprit. TCP will generally only work on LAN or exposed public IP's if you connect to a direct IP, so that one might be difficult to trigger

Retested the denied message on the main build with the same user and the same issue was exhibited. So I'm confident the issue is not related to the prerelease.

**Comment by ohzee00** [comment]
I only did some slight testing and all of it was in Jack's headless but I did a temp ban on myself (via the Resonite bot) and successfully was booted from the session _and_ was denied joining til I unbanned myself. 

I did not notice any issue with streams and they remained uninterrupted throughout the session even with people joining.

Honestly pretty impressed how there seems to be no apparent issues despite being the first prerelease of an entire join revamp! Of course, perhaps having people do self hosted sessions compared to headless might be a good test.

**Reply by Frooxius** [reply]
Thank you for testing and sharing!

**Comment by JackTheFoxOtter** [comment]
I've tested my headless and graphical client on the prerelease build yesterday, and everything seemed to work mostly fine. We played Cloudscape Harvest with others, later more people joined and everything from editing to sharing assets and streaming audio was tested - all apparently normal.

Earlier in the session EpicEaston encountered pretty severe packet loss, that was however corrected by adjusting the window size and re-joining the session. It's also not the first time she had issues on my headless, so I'm not sure I would put that on the networking changes. It did promt me to look at my own packet loss though, and I've noticed a slight discrepency.

My headless is in a data center and I'm running on residential fibreoptics, no wireless is used, and yet I noticed that my client was reporting slight packet loss in the prerelease. I've never encountered packet loss on my own headless, and sure enough after switching back to the main build without changing any networking setups, the packet loss is gone again.

Prerelease session:
![20240929155458_1](https://github.com/user-attachments/assets/67b280a5-57c3-49e1-b92b-035930801331)
[J4-C - 2024.9.28.1132 - 2024-09-29 15_41_00.log](https://github.com/user-attachments/files/17179504/J4-C.-.2024.9.28.1132.-.2024-09-29.15_41_00.log)

Session on main build:
![20240929160226_1](https://github.com/user-attachments/assets/6329c63f-d376-483a-9d62-5c7254f6e765)
[J4-C - 2024.9.16.1105 - 2024-09-29 15_58_53.log](https://github.com/user-attachments/files/17179509/J4-C.-.2024.9.16.1105.-.2024-09-29.15_58_53.log)

So something seems to be off here, even though this amount of packet loss likely won't affect anything.

**Reply by JackTheFoxOtter** [reply]
I did some more testing, the packet loss seems completely unrelated to the prerelease version, so this is a non-issue!

**Comment by ModernBalloonie** [comment]
Small question. Is this supposed to be greyed out? 
![image](https://github.com/user-attachments/assets/e530d9af-b35f-4805-aadd-a184ce58c504)

**Reply by ohzee00** [reply]
Yes, you did not set a VerifyJoinRequest node to handle it so you can not turn it on til you do.
![image](https://github.com/user-attachments/assets/0206027e-a50c-4297-889c-effc83c54441)

**Comment by Frooxius** [comment]
Thanks everyone for testing! Since this is pretty much without any major issues, I'm going to merge this in and publish a build today!

**Comment by Readun** [comment]
This might be an **edge case**, but I ran into an Issue with audio crackling way more.
I highly suspect it is my ISPs routing (Telekom) to the specific headless.  (Headless: Netherlands, Me: Germany, close to NL's border)
With Tracert I was not able to verify any annomalys compared to other people that live even futher away...

Ive been fine in other sessions and even with over 40 People on an US headless, but in this case, I got major audio crackling and from the looks of it also updating lag.
When I checked my Ping, its been around ~ 5ms. Same goes for the packetloss being some 0.000000X %.

Is there anything I can do to analyze this futher?

https://youtu.be/-dggd358a_g

**Reply by Frooxius** [reply]
Can you check packet loss on the OpusStream? You'll need to have this checked by the users who hear the crackling.

**Reply by Readun** [reply]
Its just me so far, with other people Voices.

**Reply by Frooxius** [reply]
Oh, can you check what % packet loss you see on their OpusStreams?

**Reply by Readun** [reply]
Oohh yes... 7% Packet loss only to the voice for some reason & only for me, other people have none.

PS: It -must- have been way more earlier, now the voices getting better to understand again with that 7%.
PS 2: At least I have a way to know where to check now. It seemed to went down so still highly suspect my ISP sadly with that specific connection to the host...


![2024-10-02 22 03 03](https://github.com/user-attachments/assets/17329093-106b-41d0-8600-e5b3ffce2ea2)

**Reply by Frooxius** [reply]
Does it happen with only this specific host? It sounds more like a networking issue to me. 

But yeah 7 % is quite a bit.

**Reply by Readun** [reply]
I havent expirienced it anywhere else -voice- wise.
I did experienced it with Smoin's Music stream only in the workout yesterday, but had no way to troubleshoot in that moment.

It might have been way higher than 7%, when I took the video probably....
